python/dolfin/function/expression.py

```diff
@@ -112,4 +114,11 @@ def __init__(self,
         shape: tuple
             Value shape.
         """
-        super().__init__(eval_func, shape)
+        self._eval_func = eval_func
```

Comment:
Add comment 'Hold reference to eval function to avoid premature garbage collection'.